File: src/main/java/com/github/dylmeadows/eontimer/model/Gen4TimerModel.java
package com.github.dylmeadows.eontimer.model;

import javafx.beans.property.IntegerProperty;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleIntegerProperty;
import javafx.beans.property.SimpleObjectProperty;

@SuppressWarnings({"unused", "WeakerAccess"})
public class Gen4TimerModel {

    private final ObjectProperty<Gen4TimerMode> mode = new SimpleObjectProperty<>(Gen4TimerConstants.DEFAULT_MODE);
    private final IntegerProperty calibratedDelay = new SimpleIntegerProperty(Gen4TimerConstants.DEFAULT_CALIBRATED_DELAY);
    private final IntegerProperty calibratedSecond = new SimpleIntegerProperty(Gen4TimerConstants.DEFAULT_CALIBRATED_SECOND);
    private final IntegerProperty targetDelay = new SimpleIntegerProperty(Gen4TimerConstants.DEFAULT_TARGET_DELAY);
    private final IntegerProperty targetSecond = new SimpleIntegerProperty(Gen4TimerConstants.DEFAULT_TARGET_SECOND);
    private final transient IntegerProperty delayHit = new SimpleIntegerProperty();

    public Gen4TimerMode getMode() {
        return mode.get();
    }

    public ObjectProperty<Gen4TimerMode> modeProperty() {
        return mode;
    }

    public void setMode(Gen4TimerMode mode) {
        this.mode.set(mode);
    }

    public int getCalibratedDelay() {
        return calibratedDelay.get();
    }

    public IntegerProperty calibratedDelayProperty() {
        return calibratedDelay;
    }

    public void setCalibratedDelay(int calibratedDelay) {
        this.calibratedDelay.set(calibratedDelay);
    }

    public int getCalibratedSecond() {
        return calibratedSecond.get();
    }

    public IntegerProperty calibratedSecondProperty() {
        return calibratedSecond;
    }

    public void setCalibratedSecond(int calibratedSecond) {
        this.calibratedSecond.set(calibratedSecond);
    }

    public int getTargetDelay() {
        return targetDelay.get();
    }

    public IntegerProperty targetDelayProperty() {
        return targetDelay;
    }

    public void setTargetDelay(int targetDelay) {
        this.targetDelay.set(targetDelay);
    }

    public int getTargetSecond() {
        return targetSecond.get();
    }

    public IntegerProperty targetSecondProperty() {
        return targetSecond;
    }

    public void setTargetSecond(int targetSecond) {
        this.targetSecond.set(targetSecond);
    }

    public int getDelayHit() {
        return delayHit.get();
    }

    public IntegerProperty delayHitProperty() {
        return delayHit;
    }

    public void setDelayHit(int delayHit) {
        this.delayHit.set(delayHit);
    }
}

//    calculatedCalibration = new ReadOnlyIntegerWrapper();
//    calculatedCalibration.bind(Bindings.createIntegerBinding(this::calculateCalibration,
//            calibratedDelay, calibratedSecond, consoleProperty()));
//
//    private int calculateCalibration() {
//        return CalibrationUtils.createCalibration(
//                getCalibratedDelay(),
//                getCalibratedSecond(),
//                getConsole());
//    }

File: src/models/timers/Gen4TimerModel.cpp
//
// Created by Dylan Meadows on 2020-03-17.
//

#include <QtCore>
#include "Gen4TimerModel.h"

namespace model::timer {
    namespace Gen4Fields {
        const char *CALIBRATED_DELAY = "gen4/calibratedDelay";
        const char *CALIBRATED_SECOND = "gen4/calibratedSecond";
        const char *TARGET_DELAY = "gen4/targetDelay";
        const char *TARGET_SECOND = "gen4/targetSecond";

        namespace Defaults {
            const int CALIBRATED_DELAY = 500;
            const int CALIBRATED_SECOND = 14;
            const int TARGET_DELAY = 600;
            const int TARGET_SECOND = 50;
        }
    }

    Gen4TimerModel::Gen4TimerModel(QSettings *settings, QObject *parent)
        : QObject(parent) {
        calibratedDelay = settings->value(Gen4Fields::CALIBRATED_DELAY,
                                          Gen4Fields::Defaults::CALIBRATED_DELAY).toInt();
        calibratedSecond = settings->value(Gen4Fields::CALIBRATED_SECOND,
                                           Gen4Fields::Defaults::CALIBRATED_SECOND).toInt();
        targetDelay = settings->value(Gen4Fields::TARGET_DELAY,
                                      Gen4Fields::Defaults::TARGET_DELAY).toInt();
        targetSecond = settings->value(Gen4Fields::TARGET_SECOND,
                                       Gen4Fields::Defaults::TARGET_SECOND).toInt();
    }

    void Gen4TimerModel::sync(QSettings *settings) const {
        settings->setValue(Gen4Fields::CALIBRATED_DELAY, calibratedDelay);
        settings->setValue(Gen4Fields::CALIBRATED_SECOND, calibratedSecond);
        settings->setValue(Gen4Fields::TARGET_DELAY, targetDelay);
        settings->setValue(Gen4Fields::TARGET_SECOND, targetSecond);
    }

    int Gen4TimerModel::getCalibratedDelay() const {
        return calibratedDelay;
    }

    void Gen4TimerModel::setCalibratedDelay(const int calibratedDelay) {
        if (this->calibratedDelay != calibratedDelay) {
            this->calibratedDelay = calibratedDelay;
            emit calibratedDelayChanged(calibratedDelay);
        }
    }

    int Gen4TimerModel::getCalibratedSecond() const {
        return calibratedSecond;
    }

    void Gen4TimerModel::setCalibratedSecond(const int calibratedSecond) {
        if (this->calibratedSecond != calibratedSecond) {
            this->calibratedSecond = calibratedSecond;
            emit calibratedSecondChanged(calibratedSecond);
        }
    }

    int Gen4TimerModel::getTargetDelay() const {
        return targetDelay;
    }

    void Gen4TimerModel::setTargetDelay(const int targetDelay) {
        if (this->targetDelay != targetDelay) {
            this->targetDelay = targetDelay;
            emit targetDelayChanged(targetDelay);
        }
    }

    int Gen4TimerModel::getTargetSecond() const {
        return targetSecond;
    }

    void Gen4TimerModel::setTargetSecond(const int targetSecond) {
        if (this->targetSecond != targetSecond) {
            this->targetSecond = targetSecond;
            emit targetSecondChanged(targetSecond);
        }
    }

    int Gen4TimerModel::getDelayHit() const {
        return delayHit;
    }

    void Gen4TimerModel::setDelayHit(int delayHit) {
        this->delayHit = delayHit;
    }
}
